# T05 Motor Trend Car Road Tests

|                |   |
:----------------|---|
| **Nombre**     |  Jesús Alejandro García García |
| **Fecha**      |  18-09-25 |
| **Expediente** |  751749 |

## 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

Calcula el R2 e interpreta los signos de los betas.

Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.


In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
import seaborn as sns
from scipy import stats
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge

In [2]:
df=pd.read_excel('Motor Trend Car Road Tests.xlsx')
df.head()

model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [3]:
df.drop(columns=['model'], inplace=True)

### Calcula el R2 e interpreta los signos de los betas.


In [4]:
cols=df.columns
cols=cols.drop('mpg')

x_list = []

for col in cols:
    x_list.append(df[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = df["mpg"]

scaler=StandardScaler().fit(X)
Xscaled=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xscaled))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        13:19:35   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.468     42.884      0.000      19.116      21.065
x1            -0.1959      1.837     -0.107      0.916      -4.016       3.624
x2             1.6267      2.178      0.747      0.463      -2.903       6.157
x3            -1.4497      1.469     -0.987      0.335      -4.505       1.605
x4             0.4142      0.861      0.481      0.635      -1.376       2.204
x5            -3.5780      1.824     -1.961      0.063      -7.372       0.216
x6             1.4440      1.285      1.123      0.274      -1.229       4.117
x7             0.1576      1.044      0.151      0.881      -2.013       2.329
x8             1.2378      1.010      1.225      0.234      -0.863       3.338
x9             0.4760      1.084      0.439      0.665      -1.779       2.731
x10           -0.3170      1.318     -0.241      0.812      -3.057       2.423
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                         15.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4,random_state=137)
scaler=StandardScaler().fit(X_train)
Xscaled_train=scaler.transform(X_train)

x=np.ones((len(X_train),1))
X_e_train=np.hstack((x,Xscaled_train))

ols=sm.OLS(y_train,X_e_train)
results=ols.fit()
results.summary()

c:\ProgramData\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Wed, 17 Sep 2025   Prob (F-statistic):              0.242
Time:                        13:19:36   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1            40.7571     54.010      0.755      0.588    -645.499     727.013
x2            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x3           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x4            14.1135     13.469      1.048      0.485    -157.022     185.249
x5            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x6             1.5727      3.475      0.453      0.729     -42.583      45.729
x7            32.4926     41.262      0.787      0.575    -491.795     556.780
x8            36.1394     42.669      0.847      0.553    -506.022     578.301
x9           -38.3551     43.625     -0.879      0.541    -592.668     515.958
x10           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                         382.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
Xscaled_test = scaler.transform(X_test)
x=np.ones((len(X_test),1))
X_e_test=np.hstack((x,Xscaled_test))

y_pred=results.predict(X_e_test)
r2_score(y_test,y_pred)

-54.071785204188224

### Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.


In [7]:
ridge = Ridge(alpha=10)
ridge.fit(X_e_train, y_train)
r2_train = r2_score(y_train, ridge.predict(X_e_train))
r2_test = r2_score(y_test, ridge.predict(X_e_test))
print(f'R2 Train: {r2_train}, R2 Test: {r2_test}')

R2 Train: 0.8637598356203311, R2 Test: 0.7632502786243675


Después de probar con varios alphas, encontré que 10 me daba el mejor valor posible, equilibrando el mejor resultado posible para ambas r2. Este fue una mejora exponencial al modelo que no tenía una regulación.

## 1.2 Repite el ejercicio anterior usando 'qsec' como salida.



#### Regresion clasica

In [8]:
cols=df.columns
cols=cols.drop('qsec')

x_list = []

for col in cols:
    x_list.append(df[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = df["qsec"]

scaler=StandardScaler().fit(X)
Xscaled=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xscaled))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        13:19:36   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.136    131.375      0.000      17.566      18.131
x1             0.4096      0.365      1.123      0.274      -0.349       1.168
x2            -0.6375      0.514     -1.239      0.229      -1.707       0.432
x3            -0.9150      0.608     -1.505      0.147      -2.180       0.350
x4            -0.1054      0.435     -0.242      0.811      -1.010       0.800
x5            -0.0690      0.250     -0.275      0.786      -0.590       0.452
x6             1.4410      0.482      2.990      0.007       0.439       2.443
x7             0.4812      0.284      1.694      0.105      -0.110       1.072
x8            -0.4426      0.287     -1.540      0.139      -1.040       0.155
x9            -0.1462      0.314     -0.465      0.647      -0.800       0.507
x10           -0.4350      0.371     -1.174      0.254      -1.206       0.336
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                         15.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Train/Test

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4,random_state=137)
scaler=StandardScaler().fit(X_train)
Xscaled_train=scaler.transform(X_train)

x=np.ones((len(X_train),1))
X_e_train=np.hstack((x,Xscaled_train))

ols=sm.OLS(y_train,X_e_train)
results=ols.fit()
r2_train=results.rsquared
print(f'El r cuadrado de entrenamineto es {r2_train}')
Xscaled_test = scaler.transform(X_test)
x=np.ones((len(X_test),1))
X_e_test=np.hstack((x,Xscaled_test))

y_pred=results.predict(X_e_test)
print(f'El r cuadrado de prubea es {r2_score(y_test,y_pred)}')

El r cuadrado de entrenamineto es 0.9787621925230519
El r cuadrado de prubea es 0.37610389620004314


#### Regularización L2

In [10]:
ridge = Ridge(alpha=1)
ridge.fit(X_e_train, y_train)
r2_train = r2_score(y_train, ridge.predict(X_e_train))
r2_test = r2_score(y_test, ridge.predict(X_e_test))
print(f'R2 Train: {r2_train}, R2 Test: {r2_test}')

R2 Train: 0.9671802657602482, R2 Test: 0.6535717344645449


Después de probar con varios alphas, encontré que 1 me daba el mejor valor posible, equilibrando el mejor resultado posible para ambas r2. Este fue una mejora exponencial al modelo que no tenía una regulación.

## 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

Calcula el R2 e interpreta los signos de los betas.

Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [11]:
df_dummies=pd.get_dummies(df,columns=['cyl','gear','carb'])
df_dummies.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1  21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2  22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3  21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4  18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   

   gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0    True   False   False   False   False    True   False   False  
1    True   False   False   False   False    True   False   False  
2    True   False    True   False   False   False   False   False  
3   False   False    True   False   False   False   False   False  
4   False   False   False    True   False   False   False   False

In [12]:
cols=df_dummies.columns
cols=cols.drop('mpg')

x_list = []

for col in cols:
    x_list.append(df_dummies[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = df["mpg"]

scaler=StandardScaler().fit(X)
Xscaled=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xscaled))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           0.000124
Time:                        13:19:36   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.501     40.114      0.000      19.023      21.158
x1             4.3362      3.891      1.114      0.283      -3.958      12.630
x2            -4.7580      2.661     -1.788      0.094     -10.429       0.913
x3             0.6225      1.307      0.476      0.641      -2.163       3.408
x4            -4.3624      2.445     -1.784      0.095      -9.574       0.849
x5             0.6470      1.645      0.393      0.700      -2.860       4.154
x6             0.9579      1.424      0.672      0.512      -2.078       3.994
x7             0.5953      1.578      0.377      0.711      -2.769       3.959
x8             0.3957      1.593      0.248      0.807      -3.001       3.792
x9            -0.7505      0.826     -0.909      0.378      -2.511       1.010
x10            0.2466      1.934      0.127      0.900      -3.876       4.369
x11           -0.4822      1.028     -0.469      0.646      -2.673       1.708
x12            0.0717      0.935      0.077      0.940      -1.920       2.064
x13            0.5672      0.827      0.685      0.503      -1.196       2.331
x14           -0.3513      1.056     -0.333      0.744      -2.602       1.899
x15           -0.8478      0.803     -1.055      0.308      -2.560       0.864
x16            0.6267      0.870      0.721      0.482      -1.227       2.480
x17            0.1120      1.009      0.111      0.913      -2.039       2.263
x18            0.6312      0.803      0.786      0.444      -1.081       2.344
x19            1.1137      1.130      0.986      0.340      -1.295       3.522
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     2.77e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.13e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#### Train/test

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4,random_state=137)
scaler=StandardScaler().fit(X_train)
Xscaled_train=scaler.transform(X_train)

x=np.ones((len(X_train),1))
X_e_train=np.hstack((x,Xscaled_train))

ols=sm.OLS(y_train,X_e_train)
results=ols.fit()
r2_train=results.rsquared
print(f'El r cuadrado de entrenamineto es {r2_train}')
Xscaled_test = scaler.transform(X_test)
x=np.ones((len(X_test),1))
X_e_test=np.hstack((x,Xscaled_test))

y_pred=results.predict(X_e_test)
print(f'El r cuadrado de prueba es {r2_score(y_test,y_pred)}')

El r cuadrado de entrenamineto es 1.0
El r cuadrado de prueba es 0.12691935910997576


## 2.2 Repite el ejercicio anterior usando 'qsec' como salida.



In [14]:
cols=df_dummies.columns
cols=cols.drop('qsec')

x_list = []

for col in cols:
    x_list.append(df_dummies[col].values.reshape(-1, 1))


X = np.hstack(x_list)
y = df["qsec"]

scaler=StandardScaler().fit(X)
Xscaled=scaler.transform(X)

x=np.ones((len(X),1))
X_e=np.hstack((x,Xscaled))

ols=sm.OLS(y,X_e)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        13:19:36   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.138    129.771      0.000      17.556      18.142
x1             0.1646      0.418      0.393      0.700      -0.727       1.057
x2             0.4308      1.106      0.389      0.703      -1.927       2.789
x3            -0.1394      0.804     -0.173      0.865      -1.853       1.574
x4             0.0568      0.361      0.157      0.877      -0.713       0.827
x5             0.7805      0.711      1.097      0.290      -0.735       2.297
x6             0.1318      0.396      0.333      0.744      -0.711       0.975
x7            -0.8321      0.379     -2.197      0.044      -1.639      -0.025
x8             0.6792      0.402      1.690      0.112      -0.177       1.536
x9             0.1346      0.230      0.584      0.568      -0.357       0.626
x10           -0.7624      0.494     -1.544      0.143      -1.815       0.290
x11           -0.2728      0.275     -0.990      0.338      -0.860       0.314
x12            0.3804      0.237      1.604      0.130      -0.125       0.886
x13           -0.1323      0.228     -0.580      0.571      -0.619       0.354
x14            0.3905      0.273      1.430      0.173      -0.191       0.972
x15            0.0511      0.228      0.224      0.826      -0.435       0.538
x16            0.2085      0.237      0.880      0.393      -0.296       0.713
x17           -0.4646      0.250     -1.859      0.083      -0.997       0.068
x18           -0.1080      0.223     -0.483      0.636      -0.584       0.368
x19           -0.0675      0.320     -0.211      0.836      -0.749       0.614
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     2.48e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 4.19e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4,random_state=137)
scaler=StandardScaler().fit(X_train)
Xscaled_train=scaler.transform(X_train)

x=np.ones((len(X_train),1))
X_e_train=np.hstack((x,Xscaled_train))

ols=sm.OLS(y_train,X_e_train)
results=ols.fit()
r2_train=results.rsquared
print(f'El r cuadrado de entrenamineto es {r2_train}')
Xscaled_test = scaler.transform(X_test)
x=np.ones((len(X_test),1))
X_e_test=np.hstack((x,Xscaled_test))

y_pred=results.predict(X_e_test)
print(f'El r cuadrado de prubea es {r2_score(y_test,y_pred)}')

El r cuadrado de entrenamineto es 1.0
El r cuadrado de prubea es -2.860496844108316


## 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.



Los r2 de entrenamiento y prueba sin regularización L2 del ejercicio 1.1 son de:
- El R cuadrado de entrenamiento es 0.99.
- El R cuadrado de prueba es -54.

Los r2 de entrenamiento y prueba sin regularización L2 del ejercicio 2.1 son de:
- El R cuadrado de entrenamiento es 1.0.
- El R cuadrado de prueba es 0.12691935910997576.

Ambos modelos fueron un asco y cometieron un overfitting marcadísimo, sin embargo, el ejercicio 1.1 se vio altamente beneficiado por la regularización de L2.


## 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

Los r2 de entrenamiento y prueba sin regularización L2 del ejercicio 1.2 son de:
- El R cuadrado de entrenamiento es 0.9787621925230519.
- El R cuadrado de prueba es 0.37610389620004314.

Los r2 de entrenamiento y prueba sin regularización L2 del ejercicio 2.2 son de:
- El R cuadrado de entrenamiento es 1.0.
- El R cuadrado de prueba es -2.860496844108316.

Como podemos observar, los modelos para cada uno de los ejercicios son un asco y cometen un claro overfit; sin embargo, se puede observar un claro cambio drástico en el ejercicio 2.2. Esto probablemente se pueda explicar, ya que utilizamos dummies para una variable cuantitativa, lo que nos dio varias columnas extras (varias Xs extras) y también perdió información esencial que nos daba la variable original. Todo esto causó que el modelo realizara un overfit extremo y no generalizara de manera correcta. 
